cpp/src/parquet/encoding.cc

```diff
@@ -3037,14 +3058,237 @@ class RleBooleanDecoder : public DecoderImpl, virtual public BooleanDecoder {
 // ----------------------------------------------------------------------
 // DELTA_BYTE_ARRAY
 
-class DeltaByteArrayDecoder : public DecoderImpl,
-                              virtual public TypedDecoder<ByteArrayType> {
+/// Delta Byte Array encoding also known as incremental encoding or front compression:
+/// for each element in a sequence of strings, store the prefix length of the previous
+/// entry plus the suffix.
+///
+/// This is stored as a sequence of delta-encoded prefix lengths (DELTA_BINARY_PACKED),
+/// followed by the suffixes encoded as delta length byte arrays
+/// (DELTA_LENGTH_BYTE_ARRAY).
+
+// ----------------------------------------------------------------------
+// DeltaByteArrayEncoder
+
+template <typename DType>
+class DeltaByteArrayEncoder : public EncoderImpl, virtual public TypedEncoder<DType> {
+  static constexpr std::string_view kEmpty = "";
+
  public:
-  explicit DeltaByteArrayDecoder(const ColumnDescriptor* descr,
+  using T = typename DType::c_type;
+
+  explicit DeltaByteArrayEncoder(const ColumnDescriptor* descr,
                                  MemoryPool* pool = ::arrow::default_memory_pool())
+      : EncoderImpl(descr, Encoding::DELTA_BYTE_ARRAY, pool),
+        sink_(pool),
+        prefix_length_encoder_(descr, pool),
```

Comment:
Sorry for confusion, I think we should only pass `descr` to suffix_encoder_ as it applies to ByteArrayType. But prefix_length_encoder_ is only for integer types, passing descr to it may introduce weird behavior. What about keep passing nullptr for now? Though internally the DeltaBitPackEncoder does not use the type, I am just afraid it will break in the future. WDYT?

cc @mapleFU @pitrou

Reply:
```
  EncoderImpl(const ColumnDescriptor* descr, Encoding::type encoding, MemoryPool* pool)
      : descr_(descr),
        encoding_(encoding),
        pool_(pool),
        type_length_(descr ? descr->type_length() : -1) {}
```

After checked, seems `descr` is for FLBA. I'm ok for using nullptr, and here we should add comment that why `prefix_length_encoder_` can has a null descr

Reply:
Ok, setting `suffix_encoder_(descr, pool)` and others to `..._encoder(nullptr, pool)`.